WIP: Replace lut pwf select with tabs

scottwittenburg opened the pull request:
This is coming along.  Still needed:

- [x] Better styling of tabs
- [x] text of labels better contrast
- [x] Improve look/feel of visibility checkboxes
- [x] Hook up component visibility to volume rendering
- [ ] with 2 components, second one doesn't seem to be registered until you click the tab
- [x] if only one component, don't show the tabs (and don't allow turning that comp off)

scottwittenburg commented:
@thewtex More work to do here, but the faked tabs don't look too bad given that they're actually radio buttons ;-), and the checkboxes actually work to turn on/off the contributions of the various components, but only for the slices.

thewtex commented on src/UserInterface/Image/createComponentSelector.js:
```diff
@@ -45,9 +59,21 @@ function createComponentSelector(store, imageUIGroup) {
       event.preventDefault()
       event.stopPropagation()
-      store.imageUI.selectedComponentIndex = Number(event.target.value)
+      const selIdx = Number(event.target.getAttribute('data-componentIndex'))
```
Could be cleaner here by using `dataset` exposed via [HTML5 Data Attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes).

scottwittenburg replied:
Thanks for pointing that out, it would be cleaner.  Will do.

scottwittenburg replied:
Thanks for pointing that out, I wasn't aware.  Will do.

scottwittenburg commented:
@thewtex This could go in now if you like, and I can create a new issue for the remaining unchecked item in the description.  I'm struggling to figure out why we see the behavior, and it could make sense as a separate bug anyway.

thewtex commented on src/UserInterface/index.js:
```diff
@@ -8,6 +8,7 @@ import createPointSetsUI from './createPointSetsUI'
 import emptyContainer from './emptyContainer'
 import getRootContainer from './getRootContainer'
 import checkForWebGL from './checkForWebGL'
+import { updateSliceActors, updateVolumeProps } from './updateActors'
```
Let's start organizing the rendering-related logic in `src/Rendering/`, so these functions can be exposed in `src/Rendering/index.js` instead of `src/UserInterface`.

thewtex commented on src/UserInterface/updateActors.js:
```diff
@@ -0,0 +1,46 @@
+
+function updateSliceActors(store) {
```
`updateSliceActors` -> `updateSliceProperties` for consistency with the other function and since it mostly updates the properties.

thewtex commented on src/UserInterface/updateActors.js:
```diff
@@ -0,0 +1,46 @@
+  }
+}
+
+function updateVolumeProps(store) {
```
`updateVolumeProps` -> `updateVolumeProperties` -> more specific, consistent, and not confusable with props, as in items in the scene.

thewtex commented on src/ViewerStore.js:
```diff
@@ -55,6 +55,7 @@ class ImageUIStore {
 
   lookupTableProxies = []
   piecewiseFunctionProxies = []
+  componentVisibilities = []
```
We want to make these `@observable`, then we can react to their changes.

thewtex replied:
This makes it easier to have multiple reactions for a given state change.

thewtex commented on src/UserInterface/Image/createComponentSelector.js:
```diff
@@ -47,19 +63,22 @@ function createComponentSelector(store, imageUIGroup) {
+      } else if (event.target.type === 'checkbox') {
+        const visibility = event.target.checked ? 1.0 : 0.0
+        store.imageUI.componentVisibilities[selIdx] = visibility
+        updateSliceActors(store)
+        updateVolumeProps(store)
```
We want these to calls to go into a reaction for `store.imageUI.componentVisibilities.slice()`. This way, they will always get called when the component visibilities are changes, which could happen in the user interface or from other sources, i.e. an API call that originates from Python activities.

thewtex commented on src/UserInterface/updateActors.js:
```diff
@@ -0,0 +1,46 @@
+}
+
+export {
+  updateSliceActors,
```
Exporting one function per module makes them easy to find.

scottwittenburg commented:
@thewtex I fixed the color/contrast on the selected tab as we discussed earlier today, and I think I addressed all the issues you raised in review here.  Thanks for providing feedback!
